File: sdk/python/python/agent_hooks/ctk/runner.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
"""CTK runner: load vectors, drive a harness, assert ``expect``.

The assertion engine, capability skip check, and scripted
interceptor/resolver evaluation live in the Rust core
(``_core.ctk_*``). This module keeps only:

- vector file globbing (filesystem access stays per-language),
- the orchestration loop that calls ``harness.setup/run/teardown``
  (native callbacks into the framework under test),
- ``RunRecord`` → wire-JSON marshalling for the core.

Every other language SDK's runner has the same shape.
"""

from __future__ import annotations

import asyncio
import json
import pathlib
from dataclasses import dataclass, field
from typing import Any

from agent_hooks import _core
from agent_hooks._marshal import dumps
from agent_hooks._types import EnforcementMode
from agent_hooks.composition import CompositionConfig
from agent_hooks.ctk.harness import Harness, RunRecord, Scenario
from agent_hooks.ctk.scripted import (
    RecordingInterceptor,
    ScriptedInterceptor,
    ScriptedResolver,
)


@dataclass(slots=True)
class VectorResult:
    id: str
    title: str
    status: str  # "pass" | "fail" | "skip"
    detail: str = ""
    failures: list[str] = field(default_factory=list)


def load_vectors(directory: str | pathlib.Path | None = None) -> list[dict[str, Any]]:
    """Load CTK vectors; default = the set vendored inside the wheel.

    Raises rather than returning an empty list: a runner fed zero
    vectors reports 100% pass — a false conformance signal (§13.2).
    """
    if directory is None:
        from importlib.resources import files

        root = files("agent_hooks.ctk") / "vectors"
        vectors = [
            json.loads(f.read_text(encoding="utf-8"))
            for f in sorted(root.iterdir(), key=lambda f: f.name)
            if f.name.startswith("AH-CTK-") and f.name.endswith(".json")
        ]
    else:
        d = pathlib.Path(directory)
        vectors = [
            json.loads(f.read_text(encoding="utf-8")) for f in sorted(d.glob("AH-CTK-*.json"))
        ]
    if not vectors:
        raise FileNotFoundError(
            f"no AH-CTK-*.json vectors found in {directory or 'the packaged set'} — "
            "an empty vector set would report vacuous conformance"
        )
    return vectors


def _run_record_to_wire(rr: RunRecord) -> str:
    return dumps(
        {
            "outcome": rr.outcome.value,
            "final_output": rr.final_output,
            "tool_invocations": rr.tool_invocations,
            "error": rr.error,
            "identities": [{"input_identity": i, "enforced_identity": e} for i, e in rr.identities],
            "records": rr.records,
        }
    )


async def run_vector(harness: Harness, vector: dict[str, Any]) -> VectorResult:
    vid, title = vector["id"], vector["title"]
    vector_json = dumps(vector)

    caps_json = dumps(sorted(c.value for c in harness.capabilities))
    skip = json.loads(_core.ctk_should_skip(vector_json, caps_json))
    if skip is not None:
        return VectorResult(vid, title, "skip", detail=skip)

    scenario = Scenario.from_wire(vector["scenario"])
    # Multi-interceptor vectors (§7.4 fold-through) use interceptor_scripts;
    # single-interceptor vectors use interceptor_script. Only the FIRST
    # interceptor records: expect.interceptions describes each emission as
    # the first-registered interceptor saw it.
    scripts = vector.get("interceptor_scripts")
    if scripts is None:
        scripts = [vector["interceptor_script"]]
    first = RecordingInterceptor(ScriptedInterceptor(scripts[0])) if scripts else None
    interceptors: list[Any] = [first] if first else []
    interceptors += [ScriptedInterceptor(s) for s in scripts[1:]]
    approval = vector.get("approval_script")
    resolver = ScriptedResolver(approval) if approval else None
    mode = EnforcementMode(vector.get("mode", "enforce"))
    # §13.2: composition vectors carry the profile/knobs they apply to;
    # absent means the pre-P-003 default (§7.2).
    composition = CompositionConfig.from_wire(vector.get("composition"))
    # §10.1: absent → the default provider; explicit null → unbound.
    identity_provider = vector.get("identity_provider", "jcs-sha256")
    redact_for_approval = list(vector.get("redact_for_approval", []))

    harness.setup(
        scenario,
        interceptors,
        resolver,
        mode,
        composition,
        identity_provider,
        redact_for_approval,
    )
    try:
        rr = await harness.run()
    except Exception as e:  # noqa: BLE001
        return VectorResult(vid, title, "fail", failures=[f"harness.run raised: {e!r}"])
    finally:
        harness.teardown()

    result = json.loads(
        _core.ctk_assert(
            vector_json,
            dumps(first.recorded if first else []),
            _run_record_to_wire(rr),
        )
    )
    return VectorResult(
        id=result["id"],
        title=result["title"],
        status=result["status"],
        detail=result.get("detail", ""),
        failures=result.get("failures", []),
    )


def run_vectors(harness_factory: Any, vectors: list[dict[str, Any]]) -> list[VectorResult]:
    """Run all vectors against a fresh harness instance per vector."""

    async def _go() -> list[VectorResult]:
        out = []
        for v in vectors:
            out.append(await run_vector(harness_factory(), v))
        return out

    return asyncio.run(_go())

File: sdk/go/conformance/runner.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package conformance

// CTK runner: load vectors, drive a harness, assert expect.
//
// Assertion engine, capability skip check, and scripted
// interceptor/resolver evaluation live in the Rust core via
// agenthooks.Ctk*. This file keeps only vector globbing, the
// orchestration loop that calls Harness.Setup/Run/Teardown, and
// RunRecord → wire-JSON marshalling. See conformance/RUNNER.md.

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"

	"github.com/responsibleai/agent-hooks/sdk/go/agenthooks"
)

// VectorResult is the outcome of one vector run.
type VectorResult = agenthooks.CtkVectorResult

// LoadVectors globs conformance/vectors/AH-CTK-*.json under dir in
// sorted order.
func LoadVectors(dir string) ([]map[string]any, error) {
	paths, err := filepath.Glob(filepath.Join(dir, "AH-CTK-*.json"))
	if err != nil {
		return nil, err
	}
	if len(paths) == 0 {
		// A runner fed zero vectors reports 100% pass — a false
		// conformance signal (§13.2). Fail loudly instead.
		return nil, fmt.Errorf("no AH-CTK-*.json vectors found in %s", dir)
	}
	sort.Strings(paths)
	var out []map[string]any
	for _, p := range paths {
		b, err := os.ReadFile(p)
		if err != nil {
			return nil, err
		}
		// UseNumber: default decoding into any is float64, which would
		// silently round int64_json vectors' >2^53 integers at load —
		// the very corruption AH-CTK-090 exists to catch (§4.4).
		// json.Number round-trips the literal losslessly.
		dec := json.NewDecoder(bytes.NewReader(b))
		dec.UseNumber()
		var v map[string]any
		if err := dec.Decode(&v); err != nil {
			return nil, fmt.Errorf("%s: %w", p, err)
		}
		out = append(out, v)
	}
	return out, nil
}

// scriptedInterceptor wraps agenthooks.CtkScriptedIntercept; when record
// is set it also records every context it is handed.
type scriptedInterceptor struct {
	rulesJSON string
	record    bool
	recorded  []agenthooks.AgentContext
}

func (s *scriptedInterceptor) Intercept(_ context.Context, actx agenthooks.AgentContext) (agenthooks.Verdict, error) {
	if s.record {
		cp, err := agenthooks.DeepCopyContext(actx)
		if err != nil {
			return agenthooks.Verdict{}, err
		}
		s.recorded = append(s.recorded, cp)
	}
	return agenthooks.CtkScriptedIntercept(s.rulesJSON, actx)
}

// scriptedResolver wraps agenthooks.CtkScriptedResolve.
type scriptedResolver struct {
	rulesJSON string
}

func (s *scriptedResolver) Resolve(_ context.Context, req agenthooks.ApprovalRequest) (agenthooks.ApprovalResolution, error) {
	return agenthooks.CtkScriptedResolve(s.rulesJSON, req.Context, req.ContextIdentity)
}

func mustJSON(v any) string {
	b, err := json.Marshal(v)
	if err != nil {
		panic(err)
	}
	return string(b)
}

func runRecordToWire(rr RunRecord, postures map[string]string) string {
	invs := make([]map[string]any, len(rr.ToolInvocations))
	for i, t := range rr.ToolInvocations {
		invs[i] = map[string]any{"name": t.Name, "args": t.Args}
	}
	ids := make([]map[string]any, len(rr.Identities))
	for i, p := range rr.Identities {
		ids[i] = map[string]any{
			"input_identity":    p.InputIdentity,
			"enforced_identity": p.EnforcedIdentity,
		}
	}
	// A nil slice marshals to JSON null, which the core's RunRecord
	// deserializer rejects (serde default covers absent, not null).
	records := rr.Records
	if records == nil {
		records = []agenthooks.InterceptionRecord{}
	}
	return mustJSON(map[string]any{
		"outcome":          string(rr.Outcome),
		"final_output":     rr.FinalOutput,
		"tool_invocations": invs,
		"error":            rr.Err,
		"identities":       ids,
		"records":          records,
		// Harness *declarations* (§13.1), not observed behavior: the
		// engine selects expect.run_outcome_by_posture entries by them.
		"postures": postures,
	})
}

// RunVector drives one vector against a fresh harness instance.
func RunVector(ctx context.Context, h Harness, vector map[string]any) (VectorResult, error) {
	vectorJSON := mustJSON(vector)
	id, _ := vector["id"].(string)
	title, _ := vector["title"].(string)

	caps := make([]string, 0, len(h.Capabilities()))
	for c := range h.Capabilities() {
		caps = append(caps, string(c))
	}
	sort.Strings(caps)
	if reason, err := agenthooks.CtkShouldSkip(vectorJSON, caps); err != nil {
		return VectorResult{}, err
	} else if reason != "" {
		return VectorResult{ID: id, Title: title, Status: "skip", Detail: reason}, nil
	}

	scenRaw, _ := vector["scenario"].(map[string]any)
	scenario := scenarioFromWire(scenRaw)

	// Multi-interceptor vectors (§7.1 fold-through) use
	// interceptor_scripts; single-interceptor vectors use
	// interceptor_script. Only the FIRST interceptor records:
	// expect.interceptions describes each emission as the first-registered
	// interceptor saw it. An empty interceptor_scripts registers zero
	// interceptors (§7 fail-closed vector).
	var scripts []any
	if ss, ok := vector["interceptor_scripts"].([]any); ok {
		scripts = ss
	} else {
		scripts = []any{vector["interceptor_script"]}
	}
	var first *scriptedInterceptor
	interceptors := make([]agenthooks.Interceptor, 0, len(scripts))
	for i, s := range scripts {
		si := &scriptedInterceptor{rulesJSON: mustJSON(s), record: i == 0}
		if i == 0 {
			first = si
		}
		interceptors = append(interceptors, si)
	}

	var resolver agenthooks.ApprovalResolver
	if approval, ok := vector["approval_script"].([]any); ok && len(approval) > 0 {
		resolver = &scriptedResolver{rulesJSON: mustJSON(approval)}
	}
	mode := agenthooks.Enforce
	if m, _ := vector["mode"].(string); m != "" {
		mode = agenthooks.EnforcementMode(m)
	}
	// §13.2: composition vectors carry the profile/knobs they apply to;
	// absent means the pre-P-003 default (sequential/first_deny, stop).
	composition := agenthooks.DefaultComposition()
	if c, ok := vector["composition"].(map[string]any); ok {
		var cc agenthooks.CompositionConfig
		if err := json.Unmarshal([]byte(mustJSON(c)), &cc); err == nil && cc.Profile != "" {
			composition = cc
		}
	}

	// §10.1: absent → the default provider; explicit null → unbound.
	identityProvider := agenthooks.DefaultIdentityProvider()
	if raw, present := vector["identity_provider"]; present {
		switch {
		case raw == nil:
			identityProvider = nil
		case raw == "ctk-fault":
			// §13.2: a custom provider that fails, pinning the §10.1
			// provider-failure rule (deny context_invalid pre-dispatch).
			identityProvider = &agenthooks.IdentityProvider{
				Name: "ctk-fault",
				Compute: func(agenthooks.AgentContext) (string, error) {
					return "", errors.New("ctk scripted provider fault")
				},
			}
		}
	}

	var redactForApproval []string
	if raw, ok := vector["redact_for_approval"].([]any); ok {
		for _, p := range raw {
			if sp, ok := p.(string); ok {
				redactForApproval = append(redactForApproval, sp)
			}
		}
	}

	if err := h.Setup(scenario, interceptors, resolver, mode, composition, identityProvider,
		redactForApproval); err != nil {
		return VectorResult{ID: id, Title: title, Status: "fail",
			Failures: []string{fmt.Sprintf("harness.Setup: %v", err)}}, nil
	}
	rr, runErr := h.Run(ctx)
	h.Teardown()
	if runErr != nil {
		return VectorResult{ID: id, Title: title, Status: "fail",
			Failures: []string{fmt.Sprintf("harness.Run: %v", runErr)}}, nil
	}

	recorded := []agenthooks.AgentContext{}
	if first != nil {
		recorded = first.recorded
	}
	// §13.1 posture declaration; a Harness that does not implement the
	// optional declarer interface declares the spec default.
	posture := "continue"
	if d, ok := h.(ToolSeamHostErrorDeclarer); ok {
		posture = d.ToolSeamHostError()
	}
	postures := map[string]string{"tool_seam_host_error": posture}
	return agenthooks.CtkAssert(vectorJSON, recorded, runRecordToWire(rr, postures))
}

func scenarioFromWire(s map[string]any) Scenario {
	var sc Scenario
	if in, ok := s["input"].(map[string]any); ok {
		sc.Input = in
	}
	if tools, ok := s["tools"].([]any); ok {
		for _, t := range tools {
			if m, ok := t.(map[string]any); ok {
				sc.Tools = append(sc.Tools, m)
			}
		}
	}
	if ms, ok := s["model_script"].([]any); ok {
		for _, m := range ms {
			if mm, ok := m.(map[string]any); ok {
				sc.ModelScript = append(sc.ModelScript, mm)
			}
		}
	}
	return sc
}
